src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// CREDITS
+// * https://github.com/fortanix/rust-sgx for examples of AESM requests.
+
+use crate::protobuf::aesm_proto::{
+    Request, Request_GetQuoteExRequest, Request_GetSupportedAttKeyIDNumRequest,
+    Request_GetSupportedAttKeyIDsRequest, Request_InitQuoteExRequest, Response,
+    Response_GetQuoteExResponse, Response_InitQuoteExResponse,
+};
+
+use std::io::{Error, ErrorKind, Read, Write};
+use std::mem::size_of;
+use std::os::unix::net::UnixStream;
+use std::slice::{from_raw_parts, from_raw_parts_mut};
+
+use protobuf::Message;
+use sallyport::syscall::{SGX_QUOTE_SIZE, SGX_TI_SIZE};
+
+const AESM_SOCKET: &str = "/var/run/aesmd/aesm.socket";
+const AESM_REQUEST_TIMEOUT: u32 = 1_000_000;
+const SGX_KEY_ID_SIZE: u32 = 256;
+
+// Specifies the protobuf Request type to communicate with AESMD.
+#[derive(Debug)]
+enum ReqType {
+    AkIdNum,
+    AkId,
+    TInfo,
+    KeySize,
+    Quote,
+}
+
+impl ReqType {
+    fn set_request(
+        &self,
+        report: Option<&[u8; 432]>,
+        akid: Option<Vec<u8>>,
+        size: Option<usize>,
+    ) -> Result<Request, Error> {
+        let mut req = Request::new();
+
+        match self {
+            ReqType::AkIdNum => {
+                // removed
+                unimplemented!()
+            }
+            ReqType::AkId => {
+                // removed
+                unimplemented!()
+            }
+            ReqType::TInfo => {
+                let akid = akid.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing attestation key ID from InitQuoteExRequest",
+                    )
+                })?;
+                let size = size.ok_or_else(|| {
+                    Error::new(ErrorKind::Other, "Missing key size from InitQuoteExRequest")
+                })?;
+                let mut msg = Request_InitQuoteExRequest::new();
+                msg.set_timeout(AESM_REQUEST_TIMEOUT);
+                msg.set_b_pub_key_id(true);
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(size as u64);
```

Comment:
not sure about internals of this, but if these fields always have to be set for `Request_InitQuoteExRequest` to be valid, then they should be parameters in the constructor and these setters should be removed. (I'm fine with a debt issue if that's the case)

Reply:
I think this is not possible, as it is generated code by protobuf AFAIK. @haraldh ?

Reply:
I believe this constructor is generated by protobuf.

Reply:
Yeah, so I put this resolved. I checked the whole compilation process to verify this.